### PR TITLE
Modify Fock backend, supporting MCMC measurement

### PR DESCRIPTION
**Pull request by Jooyuza**
`Fock` backend modified and new feature added:

Parameter `is_prob` of `forward` ,`forward_fock`has been changed to:
is_prob (None or bool, optional): Choices of return type. Set ``None`` to return entire unitary matrix, 
                Set ``True`` to return probabilities for Fock basis states;
                Set ``False`` to return amplitude dictionary of final state.
                Default: ``None``
Returns:
            Union[torch.Tensor, Dict]: 
            Tensor of unitary matrix(``is_prob = None``), Dictionary of final state(``is_prob = False``),
            or Dictionary of probabilty distribution(``is_prob = True``), 
            of the photonic quantum circuit after applying the ``operators``.

`Fock` backend now supports：
1. mcmc sampling;
2. batch-friendly features: when `QumodeCircuit` is encoded with batched data, `forward` function will have a batched return and `measure` function will generate batched outcome.

Brief notebook demo:"\docs\test_fockbackend_is_prob.ipynb"

### CODE REVIEW

**Review by sansiro77 — COMMENTED** [review]
Fock后端的measure根据forward输出是unitary、dict、tensor，分成三个

File: src/deepquantum/photonic/circuit.py [hunk 1]
```diff
@@ -769,20 +909,28 @@ def _prob_func_gaussian(self, state: Any) -> torch.Tensor:
...
         """The proposal sampler for MCMC sampling."""
-        sample = self._generate_rand_sample(self.detector)
+        if self.backend == 'fock':
+            assert self.basis, 'Currently NOT supported.'
+            if self.basis:
+                all_fock_basis = self._get_all_fock_basis(self.init_state.state)
+            # else:
+            #     all_fock_basis = self._get_all_fock_basis(self.init_state.state[0])
+            sample = all_fock_basis[torch.randint(0, len(all_fock_basis), (1,))][0]
+        else:
```
Comment (sansiro77) [inline src/deepquantum/photonic/circuit.py]: `elif self.backend == 'gaussian':`

**Review by sansiro77 — COMMENTED** [review]
我对docstring和代码做了些调整，可以再检查一下，再尝试支持处理带batch的初态，然后测试的ipynb别忘了删

File: src/deepquantum/photonic/circuit.py [hunk 2]
```diff
@@ -647,71 +659,199 @@ def measure(
...
         shots: int = 1024,
         with_prob: bool = False,
-        wires: Union[int, List[int], None] = None
+        wires: Union[int, List[int], None] = None,
+        mcmc: bool = False
     ) -> Union[Dict, List[Dict]]:
         """Measure the final state for Fock backend."""
+        if isinstance(self.state, torch.Tensor):
+            if self.basis:
+                return self._measure_fock_unitary(shots, with_prob, mcmc, wires)
```
Comment (sansiro77) [inline src/deepquantum/photonic/circuit.py]: bug，需要交换mcmc和wires

Comment (Jooyuza) [reply src/deepquantum/photonic/circuit.py]: 原来如此！感觉以后paramter多了还是得写成fun(a=a, b=b)

File: src/deepquantum/photonic/circuit.py [hunk 3]
```diff
@@ -671,152 +657,132 @@ def measure(
...
-                prob_dict = {}
-                wired_prob_dict = defaultdict(list)
+                state_dict = {}
+                prob_dict = defaultdict(list)
                 for i in range(len(final_states)):
-                    final_state = FockState(state=final_states[i], nmode=self.nmode, cutoff=self.cutoff, basis=self.basis)
-                    prob_dict[final_state] = rst[i]
+                    final_state = FockState(state=final_states[i])
+                    state_dict[final_state] = rst[i]
                 for key in prob_dict.keys():
```
Comment (Jooyuza) [inline src/deepquantum/photonic/circuit.py]: 找到bug了，这里是for key in state_dict.keys()

**Review by sansiro77 — APPROVED** [review]
Good job

**Pull request merged**